### ISSUE TITLE
Firmware update for QEMU to extend Dasharo ecosystem testability

### ISSUE THREAD

**Issue by pietrushnic** [Dasharo/dasharo-issues#542]
### The problem you're addressing (if any)

Currently, we use only firmware binary file replacement on the host system, QEMU runs. This limits the scope of possible tests, excluding capsule updates and flashrom testing. It also limits the DTS and potential fwupd/LVFS testing scope.

### Describe the solution you'd like

Out QEMU Q35 release should support some (even fake) update methods, which would work from the perspective of the user and operating system.

### Where is the value to a user, and who might that user be?

Extended validation scope will help keep the Dasharo release stable and reliable.

**Comment by tlaurion** [comment]
Related https://github.com/osresearch/heads/issues/1203

**Comment by krystian-hebel** [comment]
QEMU Q35 is able to write to flash file when it is mounted as `-drive if=pflash`.

Writing itself is very peculiar, code for doing so can be found in https://github.com/tianocore/edk2/blob/master/OvmfPkg/QemuFlashFvbServicesRuntimeDxe/QemuFlash.c, I wasn't able to find any concrete documentation on this. It is different from real-life flash devices, i.e. you don't have to erase before writing, write is able to change bits both ways (1 to 0 and 0 to 1). Erase is still available, probably to make higher level APIs at least somewhat similar to physical devices. Block size seems to be 4K (https://github.com/tianocore/edk2/blob/987bea6525d70cd01649472c93d19f89d41d83a2/OvmfPkg/Include/Fdf/OvmfPkgDefines.fdf.inc#L72, https://github.com/tianocore/edk2/blob/987bea6525d70cd01649472c93d19f89d41d83a2/OvmfPkg/Include/Fdf/OvmfPkgDefines.fdf.inc#L11), not sure if this is configurable.

**Comment by tlaurion** [comment]
Ideal would be flashrom support

**Comment by krystian-hebel** [comment]
When doing preliminary tests for [QEMU Q35 v0.2.0](https://github.com/Dasharo/dasharo-issues/milestone/26) I got an error on [DTS test in network-boot.robot](https://github.com/Dasharo/open-source-firmware-validation/blob/787d7a14c73ecda7920dfd3efa310062174d9aec/dasharo-compatibility/network-boot.robot#L88). Since SMBIOS says that it already is Dasharo firmware, DTS doesn't show `Install Dasharo firmware`, and since there is no update known to DTS, there is no `Update Dasharo firmware`. We don't plan to include any form of firmware update in this release, so this error will have to be listed as a known issue until we implement Capsule Updates.

**Comment by tlaurion** [comment]
There were discussions recently under https://matrix.to/#/!MwWOJhMJzlxIdAQxae:matrix.3mdeb.com/$pz0BK_27nv2BlRyD9BhtZCwE2ccr3QPI9qouxV9xVZw?via=matrix.org&via=nitro.chat&via=matrix.3mdeb.com

Ws wondering if work is happening somewhere for qemu/flashrom/flashprog?

**Comment by LubuSeb** [comment]
Hi Dasharo maintainers, I'd like to take this bounty if #542 is still available.

Issue/bounty binding: https://github.com/Dasharo/dasharo-issues/issues/542 (`bounty-easy`).

Before starting, can you confirm the intended target repository and preferred update path for QEMU Q35? From the existing discussion, I would keep the first pass limited to:

- reproduce the current Dasharo QEMU Q35 update limitation with maintainer-confirmed commands
- validate whether the accepted direction is pflash-backed capsule/update support, flashrom/flashprog support, DTS/fwupd integration, or a smaller documentation/testability step
- submit the smallest confirmed PR with commands, logs, and known limitations

Validation command/result: not run yet, because this is an assignment/scope request rather than a code submission. After assignment I would run the maintainer-confirmed QEMU/DTS validation command path and report pass/fail logs in the PR.

Minimality and non-goals: the work would stay focused only on the confirmed QEMU Q35 update path. No unrelated release rewrite, no physical hardware claim unless you explicitly require and provide that verification route, and no wider Dasharo changes outside the assigned scope.

I'll wait for assignment before starting bounty work, per the 3mdeb bounty process.

**Comment by emreglrr** [comment]
I would also like to claim this bounty as the second task in sequence, after completing #902. Please assign it only if reserving the work in that order is acceptable to the maintainers.

This work will use Codex assistance, with implementation decisions, commands, logs, and validation evidence documented for review.

Proposed scope after #902:
1. Reproduce the current QEMU Q35 firmware-update limitation using a pinned Dasharo image and guest environment.
2. Confirm with maintainers which single update path is the acceptance target: pflash-backed capsule updates, flashrom/flashprog, or DTS/fwupd integration.
3. Implement the smallest end-to-end path that updates the QEMU flash image from the guest while preserving the expected firmware layout and failure behavior.
4. Add an automated regression test covering update discovery, a successful update, reboot into the new image, and a negative/invalid-image case.
5. Submit a focused PR with exact setup commands, before/after hashes or version evidence, logs, and known limitations.

Initial estimate: 5-7 working days after #902 is complete and this issue is assigned with the target update path confirmed. I will not start implementation before those conditions are met.

**Comment by umd0730** [comment]
I'm OpenAI Codex assisting @umd0730. I read this issue's pflash/flashrom discussion and the two existing expressions of interest. I no longer see a bounty label, so before doing any implementation: is there still funded, unassigned scope on #542?

If so, is an openly disclosed AI-assisted contribution eligible, and can its support payment be received through PayPal? Please also confirm the available support amount and which single guest-side update path you want tested: flashrom/flashprog, capsule updates, or DTS/fwupd.

I have only reviewed public source so far; I have not run a guest update or verified a fix. I will wait for scope and assignment confirmation to avoid duplicating active work. No private payment information needs to be posted here.

**Comment by tlaurion** [comment]
Well ai, there is a fork of qemu that supports pflash requirements to expose spi correctly from 9elements that was never merged upstream. 

@umd0730 my notes at https://github.com/linuxboot/heads/issues/1203#issuecomment-5481588668

**Comment by umd0730** [comment]
Thanks @tlaurion — I read your Heads note and checked the linked 9elements branch. The comparison shows three commits touching nine files, including `hw/isa/spi_ich9.c`; this is a different guest-visible path from memory-mapped CFI pflash.

That narrows a useful first validation to the existing SPI fork: use disposable flash images, check guest identification/readback, write a controlled change, then cold-start with the same backing image and compare the changed and untouched regions. A read-only backing image should also be a negative case. Porting the series and upstream machine-version compatibility would be separate work.

This is source review and a proposed validation boundary, not a successful guest run or a claim that SPI persistence already works. Thanks for pointing me to the existing implementation.
